src/config.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm config option module."""
+
+import logging
+import shutil
+import subprocess
+from pathlib import Path
+from typing import Optional
+
+import ops
+from pydantic import AnyUrl, BaseModel, ValidationError, field_validator
+
+from service import FalcoLayout
+
+logger = logging.getLogger(__name__)
+
+# Executable paths
+GIT = "/usr/bin/git"
+RSYNC = "/usr/bin/rsync"
+SSH_KEYSCAN = "/usr/bin/ssh-keyscan"
+
+# Ssh related paths
+Ssh_DIR = Path.home() / ".ssh"
+SSH_KEY_FILE = Ssh_DIR / "id_rsa_falco_setting_repo"
+KNOWN_HOSTS_FILE = Ssh_DIR / "known_hosts"
+
+# Keys to look for in the setting repo.
+# See `setting-repo` config option in `charmcraft.yaml` to learn more.
+RSYNC_CUSTOM_RULES_KEY = "rules.d"
+RSYNC_CUSTOM_CONFIGS_KEY = "config.override.d"
+
+# Supported URL schemes for setting-repo
+SUPPORTED_URL_SCHEMES = "git+ssh"
+
+
+class RsyncError(Exception):
+    """Exception raised when rsync fails."""
+
+
+class GitCloneError(Exception):
+    """Exception raised when git clone fails."""
+
+
+class SshKeyScanError(Exception):
+    """Exception raised when Ssh keyscan fails."""
+
+
+class SshKeyWriteError(Exception):
+    """Exception raised when writing Ssh key fails."""
+
+
+class CharmConfig(BaseModel):
+    """Model for validating charm configuration options."""
+
+    class Config:
+        """Pydantic configuration."""
+
+        arbitrary_types_allowed = True
+
+    setting_repo: str = ""
+    setting_repo_ssh_key_id: Optional[ops.Secret] = None
+
+    @field_validator("setting_repo")
+    @classmethod
+    def validate_setting_repo(cls, v: str) -> str:
+        """Validate the setting_repo URL.
+
+        Args:
+            v (str): The setting_repo URL
+
+        Returns:
+            The validated setting_repo URL
+
+        Raises:
+            ValueError: If the URL is invalid
+        """
+        # No need to validate for empty URL
+        if not v:
+            return v
+
+        try:
+            url = AnyUrl(v)
```

Comment:
You can directly add this in the attr definition, similiar to https://github.com/canonical/wazuh-server-operator/blob/main/src/state.py#L60. It will be way more compact